hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/ImmutableSegment.java

```diff
@@ -43,4 +48,127 @@ public KeyValueScanner getKeyValueScanner() {
+      Cell c = iterator.next();
+      if (allocateFromMSLAB) {
+        // The scanner behind the iterator is doing all the elimination logic
+        // now we just copy it to the new segment (also MSLAB copy)
+        KeyValue kv = KeyValueUtil.ensureKeyValue(c);
+        Cell newKV = maybeCloneWithAllocator(kv);
+        cells[i++] = newKV;
+        // flattening = false, compaction case, counting both Heap and MetaData size
+        updateMetaInfo(c,true,!allocateFromMSLAB);
+      } else {
+        cells[i++] = c;
+        // flattening = true, flattening case, counting only MetaData size
+        updateMetaInfo(c,true,!allocateFromMSLAB);
+      }
+
+    }
+    // build the immutable CellSet
+    CellArrayMap cam = new CellArrayMap(getComparator(),cells,0,i,false);
+    return new CellSet(cam);
+  }
+
+  private CellSet createChunkBasedCellSet(
+      int numOfCells, MemStoreCompactorIterator iterator, final Configuration conf) {
+
+    int chunkSize = conf.getInt(HeapMemStoreLAB.CHUNK_SIZE_KEY, HeapMemStoreLAB.CHUNK_SIZE_DEFAULT);
+    int numOfCellsInsideChunk = chunkSize / CellChunkMap.BYTES_IN_CELL;
+    int numberOfChunks = chunkSize/numOfCellsInsideChunk;
+    HeapMemStoreLAB ms = (HeapMemStoreLAB)getMemStoreLAB();
+
+    // all Chunks must be allocated from current MSLAB
+    HeapMemStoreLAB.Chunk[] chunks = new HeapMemStoreLAB.Chunk[numberOfChunks];
+    int currentChunkIdx = 0;
+    chunks[currentChunkIdx] = ms.allocateChunk();
+    int offsetInCurentChunk = 0;
+
+    while (iterator.hasNext()) {
+      Cell c = iterator.next();
+
+      if (offsetInCurentChunk + CellChunkMap.BYTES_IN_CELL > chunkSize) {
+        // we do not consider cells bigger than chunks
+        // continue to the next chunk
+        currentChunkIdx++;
+        chunks[currentChunkIdx] = ms.allocateChunk();
+        offsetInCurentChunk = 0;
+      }
+
+      // The scanner behind the iterator is doing all the elimination logic
+      // now we just copy it to the new segment (also MSLAB copy)
+      KeyValue kv = KeyValueUtil.ensureKeyValue(c);
+      offsetInCurentChunk =
+          cloneAndReference(kv, chunks[currentChunkIdx].getData(), offsetInCurentChunk);
+    }
+
+    CellChunkMap ccm = new CellChunkMap(getComparator(),(HeapMemStoreLAB)getMemStoreLAB(),
+        chunks,0,numOfCells,chunkSize,false);
+    return new CellSet(ccm);
+  }
+
+  /**
+   * If the segment has a memory allocator the cell is being cloned to this space, and returned;
+   * otherwise the given cell is returned
+   * @return either the given cell or its clone
+   */
+  private int cloneAndReference(Cell cell, byte[] referencesByteArray, int offsetForReference) {
+    HeapMemStoreLAB ms = (HeapMemStoreLAB)getMemStoreLAB();
+    int len = KeyValueUtil.length(cell);
+    int offset = offsetForReference;
+    // we assume Cell length is not bigger than Chunk
+
+    // allocate
+    ByteRange alloc = ms.allocateBytes(len);
+    int chunkId = ms.getCurrentChunkId();
```

Comment:
this is prone to bugs
is there a way for the allocate to return both byte range and chunk id in a single object